### PR TITLE
Java: merge the @class and @interface database types and tables

### PR DESCRIPTION
**Pull request by smowton**
This will allow the extractor to emit class(id, ...) when all it knows about a class is its name, due to not having it available on the classpath. Previously it would have had to guess whether it belonged to `@class` or `@interface`, possibly introducing an inconsistency.

### CODE REVIEW

File: java/downgrades/934bf10b4bd34cf648893efcd1d0d7be9471d39f/upgrade.properties [hunk 1]
```diff
@@ -0,0 +1,5 @@
+description: Seperate class and interface tables
+compatibility: full
+classes.rel: run classes.qlo
+interfaces.rel: run interfaces.qlo
+isInterface.rel: delete
```
Comment (igfoo) [inline java/downgrades/934bf10b4bd34cf648893efcd1d0d7be9471d39f/upgrade.properties:5]: Shouldn't this also delete `classes_or_interfaces`?

File: java/ql/lib/upgrades/44d61b266bebf261cb027872646262e645efa059/upgrade.properties [hunk 2]
```diff
@@ -0,0 +1,4 @@
+description: Merge class and interface tables
+compatibility: full
+classes_or_interfaces.rel: run classes_or_interfaces.qlo
+isInterface.rel: run isInterface.qlo
```
Comment (igfoo) [inline java/ql/lib/upgrades/44d61b266bebf261cb027872646262e645efa059/upgrade.properties:4]: Shouldn't this delete `classes` and `interfaces`?

File: java/ql/lib/semmle/code/java/dispatch/internal/Unification.qll [hunk 3]
```diff
@@ -22,6 +22,17 @@ module MkUnification<unificationTarget/1 targetLeft, unificationTarget/1 targetR
     targetRight(t2) and t2.getGenericType() = g
   }
 
+  pragma[noinline]
+  private predicate unificationTargetsWithCommonSourceDecl(
+    ParameterizedType pt1, ParameterizedType pt2
+  ) {
+    exists(RefType commonSourceDecl |
+      unificationTargets(pt1, pt2) and
+      pragma[only_bind_out](pt1).getSourceDeclaration() = pragma[only_bind_out](commonSourceDecl) and
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/dispatch/internal/Unification.qll:31]: The pragma on the RHS looks superfluous to me.

**Review by aschackmull — APPROVED** [review]
One minor comment, otherwise LGTM. I presume that the up- and downgrade scripts have been tested, because AFAIR our automated testing for those doesn't quite test as much as one could hope for.

**Pull request merged**